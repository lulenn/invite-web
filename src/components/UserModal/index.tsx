import { useCallback, useState } from 'react';
import { Button, Form, Input, Modal } from 'antd';
import { IUserForm, IUserModalProps } from '../../interface';
import './index.less';
import { postUserForm } from '../../api';

const FormItem = Form.Item;

function UserModal(props: IUserModalProps): JSX.Element {
  const { visible, onOk, onCancel } = props;

  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string>('');

  const [form] = Form.useForm<IUserForm>();

  const onFinish = () => {
    const data = form.getFieldsValue();
    const { name, email } = data;
    setLoading(true);
    // request api
    postUserForm({ name, email })
      .then(x => {
        form.resetFields();
        onOk();
      })
      .catch((e: Error) => {
        console.log(e);
        setError(e.message);
      })
      .finally(() => {
        setLoading(false);
        setError('');
      });
  };

  const onValuesChange = useCallback(() => {
    setError('');
  }, []);

  const handleCancel = useCallback(() => {
    form.resetFields();
    onCancel();
  }, [onCancel]);

  return (
    <Modal visible={visible} onCancel={handleCancel} footer={null} closable={false}>
      <div className="modal-title">Request an invite</div>

      <Form name="userForm" form={form} onFinish={onFinish} onValuesChange={onValuesChange}>
        <FormItem
          name="name"
          rules={[
            {
              required: true,
              message: 'Full name is required!'
            },
            {
              min: 3,
              message: 'Full name needs to be at least 3 characters long!'
            }
          ]}
        >
          <Input placeholder="Full name" />
        </FormItem>

        <FormItem
          name="email"
          validateFirst={false}
          rules={[
            { required: true, message: 'Email is required!' },
            {
              type: 'email',
              message: 'Email is not in a validation email format!'
            }
          ]}
        >
          <Input placeholder="Email" />
        </FormItem>

        <FormItem
          name="confirmEmail"
          validateFirst={false}
          rules={[
            { required: true, message: 'Confirm email is required!' },
            ({ getFieldValue }) => ({
              validator(rule, value) {
                if (getFieldValue('email') === value) {
                  return Promise.resolve();
                }
                return Promise.reject('The confirm email does not match the email!');
              }
            })
          ]}
        >
          <Input placeholder="Confirm email" />
        </FormItem>

        <FormItem>
          <Button className="submit-btn" htmlType="submit" disabled={loading}>
            {loading ? 'Sending, please wait...' : 'Send'}
          </Button>
          <div className="error">{error || ''}</div>
        </FormItem>
      </Form>
    </Modal>
  );
}

export default UserModal;
