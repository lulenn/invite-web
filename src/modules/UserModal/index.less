.modal-title {
  font-size: 16px;
  text-align: center;
  margin: 12px 0 36px 0;
  position: relative;
  font-weight: bold;
  &::after {
    content: '';
    width: 36px;
    border-bottom: 1px solid rgba(0, 0, 0, 0.5);
    position: absolute;
    bottom: -8px;
    transform: translateX(-50%);
    left: 50%;
  }
}

.submit-btn {
  margin-top: 24px;
  width: 100%;
}

.error {
  color: red;
  text-align: center;
  margin-top: 24px;
}
