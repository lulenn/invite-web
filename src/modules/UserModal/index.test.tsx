import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import UserModal from '.';
import axiosMock from 'axios';

jest.mock('axios');

const onOk = jest.fn();
const onCancel = jest.fn();
const props = {
  visible: true,
  onOk,
  onCancel
};

describe('UserModal', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(<UserModal {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('form should validate correctly', async () => {
    render(<UserModal {...props} />);
    const btn = screen.getByText('Send');
    const nameInput = screen.getByPlaceholderText('Full name');
    const emailInput = screen.getByPlaceholderText('Email');
    const confirmEmailInput = screen.getByPlaceholderText('Confirm email');

    userEvent.click(btn);
    await waitFor(() => expect(screen.getByText('Full name is required!')).toBeInTheDocument());
    await waitFor(() => expect(screen.getByText('Email is required!')).toBeInTheDocument());
    await waitFor(() => expect(screen.getByText('Confirm email is required!')).toBeInTheDocument());

    userEvent.type(nameInput, 'aa');
    expect(nameInput).toHaveValue('aa');
    expect(screen.queryByText('Full name is required!')).toBeInTheDocument();

    userEvent.type(nameInput, 'aaa');
    expect(nameInput).toHaveValue('aaa');
    expect(screen.queryByText('Full name needs to be at least 3 characters long!')).toBeInTheDocument();

    userEvent.type(emailInput, 'invalid.email');
    expect(emailInput).toHaveValue('invalid.email');
    expect(screen.queryByText('Email is not in a validation email format!')).toBeInTheDocument();

    userEvent.type(emailInput, 'valid@xx.email.com');
    userEvent.type(confirmEmailInput, 'valid@yy.email.com');
    expect(confirmEmailInput).toHaveValue('valid@yy.email.com');
    expect(screen.queryByText('The confirm email does not match the email!')).toBeInTheDocument();
  });

  it('submit', async () => {
    render(<UserModal {...props} />);
    const btn = screen.getByText('Send');
    const nameInput = screen.getByPlaceholderText('Full name');
    const emailInput = screen.getByPlaceholderText('Email');
    const confirmEmailInput = screen.getByPlaceholderText('Confirm email');

    userEvent.type(nameInput, 'aaa');
    userEvent.type(emailInput, 'usedemail@airwallex.com');
    userEvent.type(confirmEmailInput, 'usedemail@airwallex.com');
    expect(screen.getByTestId('error')).not.toBeInTheDocument();
    // TODO:mock code here
    userEvent.click(btn);
    await waitFor(() => expect(screen.getByTestId('error')).toBeInTheDocument());

    userEvent.type(nameInput, 'bbb');
    userEvent.type(emailInput, 'valid@xx.email.com');
    userEvent.type(confirmEmailInput, 'valid@yy.email.com');
    // TODO:mock code here
    userEvent.click(btn);
    expect(onOk).toBeCalledTimes(1);
  });
});
