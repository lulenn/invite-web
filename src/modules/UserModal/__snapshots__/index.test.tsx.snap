// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserModal should take a snapshot 1`] = `<DocumentFragment />`;
